core/drivers/imx_i2c.c

```diff
@@ -0,0 +1,397 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * (c) 2020 Jorge Ramirez <jorge@foundries.io>, Foundries Ltd.
+ */
+#include <arm.h>
+#include <drivers/imx_i2c.h>
+#include <io.h>
+#include <kernel/delay.h>
+#include <kernel/generic_boot.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+#include <platform_config.h>
+#include <stdlib.h>
+#include <trace.h>
+#include <util.h>
+
+#define I2C_CLK_RATE	24000000
+
+static struct io_pa_va i2c_bus[] = {
+	{ .pa = I2C1_BASE, },
+	{ .pa = I2C2_BASE, },
+	{ .pa = I2C3_BASE, },
+};
+
+static struct imx_i2c_clk {
+	struct io_pa_va base;
+	uint32_t i2c[ARRAY_SIZE(i2c_bus)];
+} i2c_clk = {
+	.base.pa = CCM_BASE,
+	.i2c = { I2C_CLK_CGR(1), I2C_CLK_CGR(2), I2C_CLK_CGR(3), },
+};
+
+static struct imx_i2c_mux {
+	struct io_pa_va base;
+	struct imx_i2c_mux_regs {
+		uint32_t scl_mux;
+		uint32_t scl_cfg;
+		uint32_t sda_mux;
+		uint32_t sda_cfg;
+	} i2c[ARRAY_SIZE(i2c_bus)];
+} i2c_mux = {
+	.base.pa = IOMUXC_BASE,
+	.i2c = {{ .scl_mux = I2C_MUX_SCL(1), .scl_cfg = I2C_CFG_SCL(1),
+		.sda_mux = I2C_MUX_SDA(1), .sda_cfg = I2C_CFG_SDA(1), },
+	       { .scl_mux = I2C_MUX_SCL(2), .scl_cfg = I2C_CFG_SCL(2),
+		.sda_mux = I2C_MUX_SDA(2), .sda_cfg = I2C_CFG_SDA(2), },
+	       { .scl_mux = I2C_MUX_SCL(3), .scl_cfg = I2C_CFG_SCL(3),
+		.sda_mux = I2C_MUX_SDA(3), .sda_cfg = I2C_CFG_SDA(3), },},
+};
+
+#define I2DR				0x10
+#define I2SR				0x0C
+#define I2CR				0x08
+#define IFDR				0x04
+
+#define I2CR_IEN			BIT(7)
+#define I2CR_IIEN			BIT(6)
+#define I2CR_MSTA			BIT(5)
+#define I2CR_MTX			BIT(4)
+#define I2CR_TX_NO_AK			BIT(3)
+#define I2CR_RSTA			BIT(2)
+
+#define I2SR_ICF			BIT(7)
+#define I2SR_IBB			BIT(5)
+#define I2SR_IAL			BIT(4)
+#define I2SR_IIF			BIT(1)
+#define I2SR_RX_NO_AK			BIT(0)
+
+static uint8_t i2c_io_read8(uint8_t bid, uint32_t address)
+{
+	return io_read8(i2c_bus[bid].va + address);
+}
+
+static void i2c_io_write8(uint8_t bid, uint32_t address, uint8_t data)
+{
+	return io_write8(i2c_bus[bid].va + address, data);
+}
+
+static bool bus_is_idle(uint32_t sr)
+{
+	return (sr & I2SR_IBB) == 0;
+}
+
+static bool bus_is_busy(uint32_t sr)
+{
+	return !bus_is_idle(sr);
+}
+
+static bool isr_active(uint32_t sr)
+{
+	return (sr & I2SR_IIF) == I2SR_IIF;
+}
+
+/* Clock: i.MX 8M APRM Reference Manual Rev2, 08/2019, page 5281 */
```

Comment:
Could you remove this comment? This table is not SoC specific

Reply:
imx8mm comments were geared towards showing why the relevant bit of code could not be done in any other way/sequence (rather than to highlight the dependency with imx8mm). But I can see why this is confusing. will remove